File: trees/recover Binary Search Tree.java
class Solution {
    ArrayList<Integer> al = new ArrayList<Integer>();
    HashMap<Integer, TreeNode> map = new HashMap<Integer, TreeNode>();
    public void Inorder(TreeNode r){
        
        if(r == null)
              return;
        Inorder(r.left);
        al.add(r.val);
        map.put(r.val, r);
        Inorder(r.right);
    }

    public void recoverTree(TreeNode root) {
        
        Inorder(root);
        int siz = al.size();
        int f[] = new int[siz];
        int p = 0;
        for(int i=0;i<siz;i++){
            f[i] = al.get(i);
        }
        Arrays.sort(f);
        int ff[] = new int[2];
        for(int i=0;i<siz;i++){
            if(f[i] != al.get(i)){
                TreeNode fst = map.get(f[i]);
                fst.val = al.get(i);
                 TreeNode snd = map.get(al.get(i));
                snd.val = f[i];
                break;
            }
        }
        al.clear();     
        
    
    }
}

File: trees/recoverBinarySearchTree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:       
             
    # Runtime: 76 ms, faster than 86.88% of Python3 online submissions for Recover Binary Search Tree.
    # Memory Usage: 14.2 MB, less than 65.76% of Python3 online submissions for Recover Binary Search Tree.
                
    def recoverTree(self, root):
        """
        Do not return anything, modify root in-place instead.
        
        """
        
        self.first = None
        self.second = None
        self.prev = None
        
        def traverse(root):
            
            if root:
                
                traverse(root.left)
                
                if self.prev :
                    if self.first==None  and self.prev.val > root.val:
                        
                        self.first = self.prev
                        self.second = root
                        
                    elif self.first and self.prev.val > root.val:
                        
                        self.second = root
                        
                self.prev = root
                traverse(root.right)
        
        traverse(root)
        self.first.val , self.second.val = self.second.val , self.first.val
        
        
        
        